In [1]:
import pyspark
from pyspark.sql import SparkSession
from pyspark.sql import types

### Load DF from parquet

In [2]:
spark = SparkSession.builder \
    .master("local[12]") \
    .appName('hw-03') \
    .getOrCreate()

23/03/04 08:55:36 WARN Utils: Your hostname, FRANCIS-PC resolves to a loopback address: 127.0.1.1; using 172.31.24.92 instead (on interface eth0)
23/03/04 08:55:36 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address


Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


23/03/04 08:55:37 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
23/03/04 08:55:38 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.


In [3]:
df = spark.read \
    .option("header", "true") \
    .parquet('/home/fromio/cursos/de-zoom-camp/week5/parq_files/')


In [4]:
df.show(5)

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B02510|2021-06-04 08:13:36|2021-06-04 08:47:58|          91|          65|      N|                  null|
|              B02875|2021-06-25 01:52:01|2021-06-25 02:14:51|         132|         255|      N|                B02875|
|              B02875|2021-06-25 03:34:13|2021-06-25 03:42:22|          75|         263|      N|                B02875|
|              B02866|2021-06-05 09:21:57|2021-06-05 09:35:38|         233|         186|      N|                B02866|
|              B02876|2021-06-01 01:26:49|2021-06-01 01:33:02|         182|         212|      N|                B02876|
+--------------------+------------------

In [5]:
df.count()

14961892

In [6]:
df.printSchema()

root
 |-- dispatching_base_num: string (nullable = true)
 |-- pickup_datetime: timestamp (nullable = true)
 |-- dropoff_datetime: timestamp (nullable = true)
 |-- PULocationID: integer (nullable = true)
 |-- DOLocationID: integer (nullable = true)
 |-- SR_Flag: string (nullable = true)
 |-- Affiliated_base_number: string (nullable = true)



### Question 3

In [16]:
# using dataframe methods
df.filter(df.pickup_datetime >= '2021-06-15').filter(df.pickup_datetime < '2021-06-16').count()

452470

In [17]:
# with SparkQL
df.count()

14961892

In [19]:
df.createOrReplaceTempView("fhvhv_2021_06")

In [33]:
spark.sql("""SELECT *
               FROM fhvhv_2021_06 """) \
     .printSchema()

root
 |-- dispatching_base_num: string (nullable = true)
 |-- pickup_datetime: timestamp (nullable = true)
 |-- dropoff_datetime: timestamp (nullable = true)
 |-- PULocationID: integer (nullable = true)
 |-- DOLocationID: integer (nullable = true)
 |-- SR_Flag: string (nullable = true)
 |-- Affiliated_base_number: string (nullable = true)



In [44]:
spark.sql("""SELECT COUNT(1) 
               FROM fhvhv_2021_06
              WHERE DAY(pickup_datetime) = 15
          """).show()

+--------+
|count(1)|
+--------+
|  452470|
+--------+



### Question 4
Longest trip for each day

Now calculate the duration for each trip.<br>
How long was the longest trip in Hours?

In [53]:
from pyspark.sql import functions as F

In [124]:
df \
   .withColumn("duration_sec", df.dropoff_datetime.cast("long") - df.pickup_datetime.cast("long") ) \
   .groupBy('pickup_datetime') \
   .max('duration_sec') \
   .orderBy('max(duration_sec)', ascending=False) \
   .show()


23/03/04 11:18:34 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.
23/03/04 11:18:34 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.
23/03/04 11:18:34 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.
23/03/04 11:18:34 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.
23/03/04 11:18:34 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.
23/03/04 11:18:35 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.
23/03/04 11:18:35 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.
23/03/04 11:18:35 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.
23/03/04 11:18:35 WARN RowBasedKeyValueBatch: Calling spill() on

23/03/04 11:18:35 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.
23/03/04 11:18:35 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.
23/03/04 11:18:35 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.
23/03/04 11:18:35 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.
23/03/04 11:18:35 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.
23/03/04 11:18:35 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.
23/03/04 11:18:35 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.
23/03/04 11:18:35 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.
23/03/04 11:18:35 WARN RowBasedKeyValueBatch: Calling spill() on

+-------------------+-----------------+
|    pickup_datetime|max(duration_sec)|
+-------------------+-----------------+
|2021-06-25 13:55:41|           240764|
|2021-06-22 12:09:45|            91979|
|2021-06-27 10:32:29|            71931|
|2021-06-26 22:37:11|            65510|
|2021-06-23 20:40:43|            59281|
|2021-06-23 22:03:31|            51368|
|2021-06-24 23:11:00|            50075|
|2021-06-04 20:56:02|            42012|
|2021-06-27 07:45:19|            40917|
|2021-06-20 17:05:12|            39544|
|2021-06-01 12:25:29|            36963|
|2021-06-28 13:13:59|            35879|
|2021-06-01 12:01:46|            35879|
|2021-06-27 03:52:14|            34696|
|2021-06-18 08:50:29|            34648|
|2021-06-08 16:38:14|            34129|
|2021-06-11 23:26:20|            34098|
|2021-06-15 06:47:22|            33848|
|2021-06-25 02:32:24|            33817|
|2021-06-04 17:41:23|            33757|
+-------------------+-----------------+
only showing top 20 rows



In [123]:
spark.sql("""
WITH
     query_dur AS (
          SELECT
               dropoff_datetime - pickup_datetime AS trip_duration
          FROM
               fhvhv_2021_06
     ),
     query_extracted AS (
          SELECT
               DATE_PART ('DAY', trip_duration) AS dur_days,
               DATE_PART ('HOUR', trip_duration) AS dur_hours,
               DATE_PART ('MINUTE', trip_duration) AS dur_min,
               DATE_PART ('SECOND', trip_duration) AS dur_sec
          FROM
               query_dur
     )
SELECT
     dur_days,
     dur_hours,
     dur_min,
     dur_sec,
     dur_days * 24 + dur_hours + dur_min / 60 + dur_sec / 3600 AS total_hours
FROM
     query_extracted
ORDER BY
     total_hours DESC
LIMIT
     5
""").show()

+--------+---------+-------+---------+------------------+
|dur_days|dur_hours|dur_min|  dur_sec|       total_hours|
+--------+---------+-------+---------+------------------+
|       2|       18|     52|44.000000| 66.87888888888666|
|       1|        1|     32|59.000000|25.549722222223334|
|       0|       19|     58|51.000000|19.980833333336665|
|       0|       18|     11|50.000000|18.197222222223335|
|       0|       16|     28| 1.000000|16.466944444446664|
+--------+---------+-------+---------+------------------+



### Question 5

**A**: port 4040

http://localhost:4040/jobs/

### Question 6

**Most frequent pickup location zone**<BR>
Load the zone lookup data into a temp view in Spark<br>
Using the zone lookup data and the fhvhv June 2021 data, what is the name of the most frequent pickup location zone?

In [145]:
# downloading...
zones_url = "https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv"
local_path = "/home/fromio/cursos/de-zoom-camp/week5/taxi_zone_lookup.csv"

!wget {zones_url} -O {local_path}


--2023-03-04 11:54:12--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv
Resolving github.com (github.com)... 20.201.28.151
Connecting to github.com (github.com)|20.201.28.151|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6ea97ed0e6a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230304%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230304T145415Z&X-Amz-Expires=300&X-Amz-Signature=184f0e7e8aa102f5dac507e476dbf34f95196569baa22dbdf58d9b0e26c32094&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dtaxi_zone_lookup.csv&response-content-type=application%2Foctet-stream [following]
--2023-03-04 11:54:13--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a

In [153]:
df_zones = spark.read \
        .options(header=True) \
        .csv(local_path)

df_zones.printSchema()

root
 |-- LocationID: string (nullable = true)
 |-- Borough: string (nullable = true)
 |-- Zone: string (nullable = true)
 |-- service_zone: string (nullable = true)



In [154]:
zones_schema = types.StructType([
 types.StructField("LocationID", types.IntegerType(), True),
 types.StructField("Borough", types.StringType(), True),
 types.StructField("Zone", types.StringType(), True),
 types.StructField("service_zone", types.StringType(), True) 
 ])

df_zones = spark.read \
        .options(header=True) \
        .schema(zones_schema) \
        .csv(local_path)

df_zones.show()

+----------+-------------+--------------------+------------+
|LocationID|      Borough|                Zone|service_zone|
+----------+-------------+--------------------+------------+
|         1|          EWR|      Newark Airport|         EWR|
|         2|       Queens|         Jamaica Bay|   Boro Zone|
|         3|        Bronx|Allerton/Pelham G...|   Boro Zone|
|         4|    Manhattan|       Alphabet City| Yellow Zone|
|         5|Staten Island|       Arden Heights|   Boro Zone|
|         6|Staten Island|Arrochar/Fort Wad...|   Boro Zone|
|         7|       Queens|             Astoria|   Boro Zone|
|         8|       Queens|        Astoria Park|   Boro Zone|
|         9|       Queens|          Auburndale|   Boro Zone|
|        10|       Queens|        Baisley Park|   Boro Zone|
|        11|     Brooklyn|          Bath Beach|   Boro Zone|
|        12|    Manhattan|        Battery Park| Yellow Zone|
|        13|    Manhattan|   Battery Park City| Yellow Zone|
|        14|     Brookly

In [167]:
df_zones.createOrReplaceTempView("taxi_zones")

In [178]:
spark.sql("""
SELECT
     zp.Zone,
     zp.Borough,
     COUNT(1) AS total
FROM
     fhvhv_2021_06 trips
     JOIN taxi_zones zp ON trips.PULocationID = zp.LocationID
GROUP BY
     zp.Zone,
     zp.Borough
ORDER BY
     total DESC
LIMIT
     5
""") \
.show()

+-------------------+---------+------+
|               Zone|  Borough| total|
+-------------------+---------+------+
|Crown Heights North| Brooklyn|231279|
|       East Village|Manhattan|221244|
|        JFK Airport|   Queens|188867|
|     Bushwick South| Brooklyn|187929|
|      East New York| Brooklyn|186780|
+-------------------+---------+------+

